compiler/back_end/cpp/header_generator.py

```diff
@@ -1330,6 +1349,143 @@ def _cpp_field_name(name):
         return name
 
 
+def _is_equality_check(expression):
+    return (
+        expression.which_expression == "function"
+        and expression.function.function == ir_data.FunctionMapping.EQUALITY
+        and len(expression.function.args) == 2
+    )
+
+
+def _render_case_label(expression, ir):
+    if expression.type.which_type == "integer":
+        return _render_integer(int(expression.type.integer.modular_value))
+    elif expression.type.which_type == "enumeration":
+        # Need fully qualified enum type name
+        enum_type = expression.type.enumeration
+        cpp_enum_type = _get_fully_qualified_name(enum_type.name.canonical_name, ir)
+        return "static_cast</**/{}>({})".format(cpp_enum_type, enum_type.value)
+    else:
+        assert False, "Unsupported switch case type"
+
+
+def _get_switch_candidate(expression, ir):
+    """Returns (discriminant_expr, case_value_expr) or (None, None)."""
+    if not _is_equality_check(expression):
+        return None, None
+
+    arg0 = expression.function.args[0]
+    arg1 = expression.function.args[1]
+
+    # Check if args are integer or enumeration (ignore boolean for switch)
+    if arg0.type.which_type not in ("integer", "enumeration"):
+        return None, None
+
+    res0 = _render_expression(arg0, ir, subexpressions=None)
+    res1 = _render_expression(arg1, ir, subexpressions=None)
+
+    if res0.is_constant and not res1.is_constant:
+        return arg1, arg0
+    if res1.is_constant and not res0.is_constant:
+        return arg0, arg1
+
+    return None, None
+
+
+def _generate_optimized_ok_method_body(fields, ir, subexpressions):
+    groups = {}
+    ordered_keys = []
+
+    for field in fields:
+        discrim_expr, case_expr = _get_switch_candidate(field.existence_condition, ir)
+        if discrim_expr:
+            discrim_str = _render_expression(
+                discrim_expr, ir, subexpressions=None
+            ).rendered
+            key = "SWITCH:" + discrim_str
+            if key not in groups:
+                groups[key] = {"type": "switch", "expr": discrim_expr, "cases": []}
+                ordered_keys.append(key)
+
+            case_str = _render_case_label(case_expr, ir)
+            groups[key]["cases"].append((case_str, field))
+        else:
+            cond_res = _render_expression(
+                field.existence_condition, ir, subexpressions=None
+            )
+            key = "IF:" + cond_res.rendered
+            if key not in groups:
+                groups[key] = {
+                    "type": "if",
+                    "expr": field.existence_condition,
+                    "fields": [],
+                }
+                ordered_keys.append(key)
+            groups[key]["fields"].append(field)
+
+    lines = []
+    for key in ordered_keys:
+        group = groups[key]
+        if group["type"] == "switch":
```

Comment:
In general we maybe want to use templates for the switch and if blocks, this is a little hard to follow.